packages/protocol/lib/backward/report.ts

```diff
@@ -0,0 +1,176 @@
+// tslint:disable: max-classes-per-file
+import {
+  ASTCodeCompatibilityReport,
+  Change
+} from '@celo/protocol/lib/backward/ast-code'
+import { ASTStorageCompatibilityReport } from '@celo/protocol/lib/backward/ast-layout'
+import { categorize, Categorizer, ChangeType } from '@celo/protocol/lib/backward/categorizer'
+import { ContractVersionDelta } from '@celo/protocol/lib/backward/version'
+
+/**
+ * Value object holding all uncategorized storage and code reports.
+ */
+export class ASTReports {
+  
+  constructor(
+    public readonly code: ASTCodeCompatibilityReport,
+    public readonly storage: ASTStorageCompatibilityReport[]) {}
+
+  /**
+   * @return a new {@link ASTReports} with the same storage and code
+   * reports, excluding all contract names that match the {@param exclude}
+   * parameter.
+   */
+  excluding = (exclude: RegExp): ASTReports => {
+    const included = (contract: string): boolean => {
+      if (exclude != null) {
+        return !exclude.test(contract)
+      }
+      return true
+    }
+    const codeReport = new ASTCodeCompatibilityReport(this.code.getChanges().filter(r => included(r.getContract())))
+    const storageReports = this.storage.filter(r => included(r.contract))
+    return new ASTReports(codeReport, storageReports)
+  }
+}
+
+/**
+ * A mapping {contract name => {@link CategorizedChanges}}.
+ */
+export interface CategorizedChangesIndex {
+  [contract: string]: CategorizedChanges;
+}
+
+/**
+ * A {@link CategorizedChanges} builder pattern implementation.
+ */
+class CategorizedChangesBuilder {
+  public storage: ASTStorageCompatibilityReport[] = []
+  public major: Change[] = []
+  public minor: Change[] = []
+  public patch: Change[] = []
+  build = (): CategorizedChanges => {
+    return new CategorizedChanges(this.storage, this.major, this.minor, this.patch)
+  }
+}
+
+/**
+ * A semantic versioning categorization of a list of contract storage
+ * and code changes.
+ */
+export class CategorizedChanges {
+
+  /**
+   * @returns a new {@link CategorizedChanges} according to
+   * the {@link Categorizer} given.
+   */
+  static fromReports(
+    reports: ASTReports,
+    categorizer: Categorizer): CategorizedChanges {
+    const storage = reports.storage.filter(r => !r.compatible)
+    const c = categorize(reports.code.getChanges(), categorizer)
+    const major = c[ChangeType.Major]
+    const minor = c[ChangeType.Minor]
+    const patch = c[ChangeType.Patch]
+    return new CategorizedChanges(storage, major, minor, patch)
+  }
+
+  constructor(
+    public readonly storage: ASTStorageCompatibilityReport[],
+    public readonly major: Change[],
+    public readonly minor: Change[],
+    public readonly patch: Change[]) {}
+
+
+
+  /**
+   * @returns a mapping {contract name => {@link CategorizedChanges}}
+   */
+  explode = (): CategorizedChangesIndex => {
+    const builders: {[k: string]: CategorizedChangesBuilder} = {}
+    const builder = (k: string) => {
```

Comment:
I didn't mind much about it since it's just an internal variable for the method, but if it's not clear enough I'll rename it to 'contract' (it's actually the contract name, but I think contract will suffice)